packages/kit/src/index.tsx

```diff
@@ -1,37 +1,83 @@
-  },
-  RiseSelectField: {
-    component: SelectField,
-    validator: SelectField.validate,
-  },
-  RiseSlider: {
-    component: Slider,
-    validator: Slider.validate,
-  },
-  RiseSliderField: {
-    component: SliderField,
-    validator: SliderField.validate,
-  },
-  RiseSortableList: {
-    component: SortableList,
-    validator: SortableList.validate,
-  },
-  RiseSwitchField: {
-    component: SwitchField,
-    validator: SwitchField.validate,
-  },
+import { ReactElement } from 'react'
+import z from 'zod'
+
+function createServerComponent<T = any>(type: string, validator: (props: T) => T) {
+  return (props: T): ReactElement => ({
+    type,
+    props: validator(props),
+    key: null,
+  })
 }
+
+export type IconProps = z.infer<typeof IconSchema>
+export const IconSchema = z.object({
+  icon: z.string(),
+  size: z.number().optional(),
+})
+export const Icon = createServerComponent('RiseIcon', IconSchema.parse)
+
+export type QRCodeProps = z.infer<typeof QRCodeSchema>
+export const QRCodeSchema = z.object({
+  value: z.string().nullable().optional(),
+})
+export const QRCode = createServerComponent('RiseQRCode', QRCodeSchema.parse)
+
+export type SelectFieldProps = z.infer<typeof SelectFieldSchema>
+export const SelectFieldSchema = z.object({
+  value: z.string().nullable(),
+  id: z.string().optional(),
+  label: z.string().optional(),
+  unselectedLabel: z.string().optional().default('...'),
+  onValueChange: z.function().optional(),
+  options: z
+    .array(
+      z.object({
+        key: z.string(),
+        label: z.string(),
+      })
+    )
+    .optional(),
+})
+export const SelectField = createServerComponent('RiseSelectField', SelectFieldSchema.parse)
+
+export type SliderProps = z.infer<typeof SliderSchema>
+export const SliderSchema = z.object({
+  value: z.number(),
+  min: z.number().optional().default(0),
+  max: z.number().optional().default(100),
+  step: z.number().optional().default(1),
+  onValueChange: z.function().optional(),
+})
+export const Slider = createServerComponent('RiseSlider', SliderSchema.parse)
+
+export type SliderFieldProps = z.infer<typeof SliderFieldSchema>
+export const SliderFieldSchema = SliderSchema.extend({
+  defaultValue: z.number().optional(),
+  onValueChange: z.function().optional(),
+  label: z.string().optional(),
+})
+export const SliderField = createServerComponent('RiseSliderField', SliderFieldSchema.parse)
+
+export type SortableListItemSchema = z.infer<typeof SortableListItemSchema>
+export const SortableListItemSchema = z.object({
+  key: z.string(),
+  label: z.string(),
+  onPress: z.function().optional(),
+})
+
+export type SortableListProps = z.infer<typeof SortableListSchema>
+export const SortableListSchema = z.object({
+  footer: z.any(),
+  items: z.array(SortableListItemSchema),
+  // tbd: support this event again
+  onReorder: z.function().optional(),
+})
+export const SortableList = createServerComponent('RiseSortableList', SortableListSchema.parse)
```

Comment:
A bit of boilerplate, but it a) works and b) does not leak client code onto the server. Suggestions welcome to reduce it even further down. This needs to be improved, so people can write their own components within a single file, not two or three.